java/ql/src/semmle/code/java/security/LdapInjection.qll

```diff
@@ -0,0 +1,436 @@
+/** Provides classes to reason about LDAP injection attacks. */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.frameworks.Jndi
+import semmle.code.java.frameworks.UnboundId
+import semmle.code.java.frameworks.SpringLdap
+import semmle.code.java.frameworks.ApacheLdap
+
+/** A data flow sink for unvalidated user input that is used to construct LDAP queries. */
+abstract class LdapInjectionSink extends DataFlow::Node { }
+
+/** A sanitizer that prevents LDAP injection attacks. */
+abstract class LdapInjectionSanitizer extends DataFlow::Node { }
+
+/** Holds if the JNDI method parameter at index is susceptible to a LDAP injection attack. */
+private predicate jndiLdapInjectionSinkMethod(Method m, int index) {
+  m.getDeclaringType().getAnAncestor() instanceof TypeDirContext and
+  m.hasName("search") and
+  index in [0 .. 1]
+}
+
+/**
+ * JNDI sink for LDAP injection vulnerabilities, i.e. 1st (DN) or 2nd (filter) argument to
+ * `search` method from `DirContext`.
+ */
+private class JndiLdapInjectionSink extends LdapInjectionSink {
```

Comment:
The way that the various `LdapInjectionSink`s have been split into separate classes creates some code duplication. I suggest reintroducing the deleted `predicate ldapInjectionSinkMethod(Method m, int index)` and having just a single class `DefaultLdapInjectionSink extends LdapInjectionSink`.